src/Components/Server/src/BlazorPack/BlazorPackHubProtocolWorker.cs

```diff
@@ -34,6 +36,21 @@ protected override object DeserializeObject(ref MessagePackReader reader, Type t
                 {
                     return reader.ReadSingle();
                 }
+                else if (type == typeof(byte[][]))
+                {
+                    if (reader.IsNil)
+                    {
+                        return null;
+                    }
+
+                    var length = reader.ReadArrayHeader();
+                    var result = new List<byte[]>();
+                    for (var i = 0; i < length; i++)
+                    {
+                        result.Add(reader.ReadBytes().GetValueOrDefault().ToArray());
+                    }
+                    return result.ToArray();
```

Comment:
I notice this logic is doing all the allocations for the byte arrays up front. That might be the only possible way to do it, but I'm asking to be sure.

### Reason

There's no maximum number of entries in this `byte[][]`, other than the constraint imposed by the SignalR max message size. So as it stands, if the SignalR max message size was (say) 32KB, and if MessagePack could represent an empty array in (say) 4 bytes, then this logic could have the server perform 8000-ish heap allocations on each incoming message, whether or not there even is any JS invokable endpoint that accepts binary data.

I suppose that's not much different from a JSON payload that contains 8000 very short distinct strings, so maybe it's nothing to worry about.

### Possible alternative

If this logic returned a `List<ReadOnlySequence<byte>>` (or similar) instead, would that avoid the per-entry allocations? I guess it depends on how MessagePack's `ReadBytes` is implemented. But it we were able to do this and pass it through to the JSON deserializer, then the JSON deserializer would only need to convert the `ReadOnlySequence<byte>` into a `byte[]` in the case where it matches an actual `byte[]` property declared on a .NET type. So a mischief-maker wouldn't be able to force any more allocations than the target data structure declares (except if the target data structure contains something like a `List`).

I know it really might not make any difference, given that people will sometimes declare List-type structures on their target types. But if it's no more difficult to delay the allocations instead of doing them up front, it's worth considering.

Reply:
Update: the suggestion below on a different transport mechanism would make this concern obsolete.